File: LeetCode/3281-maximize-score-of-numbers-in-ranges/3281-maximize-score-of-numbers-in-ranges.py
class Solution:
    def maxPossibleScore(self, start: List[int], d: int) -> int:
        start.sort()
        n = len(start)
        low, high = 0, start[n - 1] + d - start[0]
        ans = 0

        def pranavjawale01(mid: int) -> bool:
            ans = start[0]
            for i in range(1, n):
                if mid + ans <= start[i] + d:
                    ans = max(ans + mid, start[i])
                else:
                    return False
            return True

        while low <= high:
            mid = low + (high - low) // 2
            if pranavjawale01(mid):
                ans = mid
                low = mid + 1
            else:
                high = mid - 1

        return ans

File: LeetCode/3281-maximize-score-of-numbers-in-ranges/3281-maximize-score-of-numbers-in-ranges.cpp
using ll = long long;

class Solution {
public:
    bool pranavjawale01(vector<int>& start, int d, int n, int mid) {
        ll ans = start[0];
        for (int i = 1; i <= n; i++) {
            if (ans + mid <= (ll)start[i] + d) {
                ans = max(ans + mid, (ll)start[i]);
            } else {
                return false;
            }
        }
        return true;
    }
    int maxPossibleScore(vector<int>& start, int d) {
        sort(start.begin(), start.end());
        int n = start.size() - 1;
        int low = 0, high = start[n] + d - start[0];
        int ans = 0;
        while (low <= high) {
            int mid = low + (high - low) / 2;
            if (pranavjawale01(start, d, n, mid)) {
                ans = mid;
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }
        return ans;
    }
};